src/mainwin.cc

```diff
@@ -1027,83 +1027,89 @@ void MainWindow::refreshDocument()
+				}
 			}
 		}
-		// If the file hasn't changed, we might still need to compile it
-		// if we haven't yet compiled the current text.
 		else {
-			auto current_doc = editor->toPlainText();
-			if (current_doc != last_compiled_doc && last_compiled_doc.size() == 0) {
+			shouldcompiletoplevel = true;
+		}
+
+		if (!shouldcompiletoplevel && this->parsed_module) {
+			auto mtime = this->parsed_module->includesChanged();
+			if (mtime > this->includes_mtime) {
+				this->includes_mtime = mtime;
 				shouldcompiletoplevel = true;
 			}
 		}
-	}
-	else {
-		shouldcompiletoplevel = true;
-	}
 
-	if (!shouldcompiletoplevel && this->parsed_module) {
-		auto mtime = this->parsed_module->includesChanged();
-		if (mtime > this->includes_mtime) {
-			this->includes_mtime = mtime;
-			shouldcompiletoplevel = true;
+		if (shouldcompiletoplevel) {
+			if (editor->isContentModified()) saveBackup();
+			compileTopLevelDocument(rebuildParameterWidget);
+			didcompile = true;
 		}
-	}
 
-	if (shouldcompiletoplevel) {
-		if (editor->isContentModified()) saveBackup();
-		compileTopLevelDocument(rebuildParameterWidget);
-		didcompile = true;
-	}
-
-	if (this->root_module) {
-		auto mtime = this->root_module->handleDependencies();
-		if (mtime > this->deps_mtime) {
-			this->deps_mtime = mtime;
-			PRINTB("Used file cache size: %d files", ModuleCache::instance()->size());
-			didcompile = true;
+		if (this->root_module) {
+			auto mtime = this->root_module->handleDependencies();
+			if (mtime > this->deps_mtime) {
+				this->deps_mtime = mtime;
+				PRINTB("Used file cache size: %d files", ModuleCache::instance()->size());
+				didcompile = true;
+			}
 		}
-	}
 
-	// If we're auto-reloading, listen for a cascade of changes by starting a timer
-	// if something changed _and_ there are any external dependencies
-	if (reload && didcompile && this->root_module) {
-		if (this->root_module->hasIncludes() ||
-				this->root_module->usesLibraries()) {
-			this->waitAfterReloadTimer->start();
-			this->procevents = false;
-			return;
+		// If we're auto-reloading, listen for a cascade of changes by starting a timer
+		// if something changed _and_ there are any external dependencies
+		if (reload && didcompile && this->root_module) {
+			if (this->root_module->hasIncludes() ||
+					this->root_module->usesLibraries()) {
+				this->waitAfterReloadTimer->start();
+				this->procevents = false;
+				return;
+			}
 		}
-	}
 
-	if (!reload && didcompile) {
-		if (!animate_panel->isVisible()) {
-			emit unhighlightLastError();
-			if (!this->root_module) {
-				emit highlightError( parser_error_pos );
+		if (!reload && didcompile) {
+			if (!animate_panel->isVisible()) {
+				emit unhighlightLastError();
+				if (!this->root_module) {
+					emit highlightError( parser_error_pos );
+				}
 			}
 		}
-	}
 
-	compileDone(didcompile | forcedone);
+		compileDone(didcompile | forcedone);
+	}catch(HardWarningException){
+		exceptionCleanup();
+	}
+	OpenSCAD::hardwarnings = false;
```

Comment:
Why are we setting + resetting this flag in various methods? Shouldn't this be a constant only set from Preferences?

Reply:
The problem is, that we generate various warnings in various places.
Specifically: Any `WARNING:` in mainwin.cc is problematic for the approach I took. (Throwing an exception on the first `WARNING:`)
Not all warnings have something to do with the SCAD File currently processed.
https://github.com/openscad/openscad/blob/15410be568ed79c172e293dc62655b287d5449a1/src/mainwin.cc#L1483
https://github.com/openscad/openscad/blob/15410be568ed79c172e293dc62655b287d5449a1/src/mainwin.cc#L1501
https://github.com/openscad/openscad/blob/15410be568ed79c172e293dc62655b287d5449a1/src/mainwin.cc#L1592
Not wanting to mess with those, I decided to simply only enable hardwarnings, when the code block is guarded by a try() catch. But now that I think about that, this is a bad idea. We do not fully lock down the UI when processing a design, so the user can open a new window. Yeah... there is no way around it. The UI-Warnings need to use a different magic word.